core/src/worker/activities.rs

```diff
@@ -145,87 +139,216 @@ impl WorkerActivityTasks {
+        stream::select_with_strategy(non_poll_stream, poller_stream, |_: &mut ()| PollNext::Left)
+            .zip(cloning_stream)
+            .map(|(res, (metrics, outstanding_tasks))| {
+                res.map(|(task, is_eager)| {
+                    Self::about_to_issue_task(outstanding_tasks, task, is_eager, metrics)
+                })
+                .map_err(|err| err.into())
+            })
+            // This map, chain, filter_map sequence is here to cancel the token when this stream ends.
+            .map(Some)
+            .chain(futures::stream::once(async move {
+                on_complete_token.cancel();
+                None
+            }))
+            .filter_map(future::ready)
+    }
+
+    /// Builds an [ActivityTask] stream for cancellation tasks from cancels delivered by the `heartbeat_manager`
+    fn make_cancel_task_stream(
+        cancels_stream: UnboundedReceiverStream<PendingActivityCancel>,
+        outstanding_tasks: Arc<DashMap<TaskToken, RemoteInFlightActInfo>>,
+    ) -> impl Stream<Item = Result<ActivityTask, PollActivityError>> {
+        let cloning_stream = stream::unfold(outstanding_tasks, |outstanding_tasks| async move {
+            Some((outstanding_tasks.clone(), outstanding_tasks))
+        });
+        cancels_stream.zip(cloning_stream).filter_map(|(next_pc, outstanding_tasks)| async move {
+            // It's possible that activity has been completed and we no longer have an
+            // outstanding activity task. This is fine because it means that we no
+            // longer need to cancel this activity, so we'll just ignore such orphaned
+            // cancellations.
+            if let Some(mut details) = outstanding_tasks.get_mut(&next_pc.task_token) {
+                if details.issued_cancel_to_lang {
+                    // Don't double-issue cancellations
+                    return None
+                }
+
+                details.issued_cancel_to_lang = true;
+                if next_pc.reason == ActivityCancelReason::NotFound {
+                    details.known_not_found = true;
+                }
+                Some(Ok(ActivityTask::cancel_from_ids(next_pc.task_token.0, next_pc.reason)))
+            } else {
+                debug!(task_token = ?next_pc.task_token, "Unknown activity task when issuing cancel");
+                // If we can't find the activity here, it's already been completed,
+                // in which case issuing a cancel again is pointless.
+                None
+            }
+        })
+    }
+
     pub(crate) fn notify_shutdown(&self) {
-        self.poller.notify_shutdown();
+        self.poller_shutdown_token.cancel();
+        self.eager_activities_semaphore.close();
     }
 
     /// Wait for all outstanding activity tasks to finish
-    pub(crate) async fn wait_all_finished(&self) {
-        while !self.outstanding_activity_tasks.is_empty() {
-            self.complete_notify.notified().await
+    async fn wait_all_finished(
```

Comment:
TODO: this seems to require a better name now that it also notifies shutdown.
Maybe `wait_drain_and_notify_heartbeat_manager`...

Reply:
I'm just worried about races if we do this synchronously.
I would agree that your approach is better if we had a single place that manages all of the state but that's not the case.

Reply:
Another option is to feed this future into the heartbeat manager.

Reply:
I think I managed to get rid of this by using `take_until` on the cancelled stream.